app/main.py

```diff
@@ -1,214 +1,115 @@
+class Knight:
+    def __init__(self, name: str, power: int, hp: int, armour: str,
+                 weapon: str, potion: str):
+        self.name = name
+        self.base_power = power
+        self.hp = hp
+        self.armour = armour or []
+        self.weapon = weapon
+        self.potion = potion
+        self.protection = self.calculate_protection()
+        self.power = self.calculate_power()
+
+    def calculate_protection(self):
+        protection = sum(a["protection"] for a in self.armour)
+        if self.potion and "protection" in self.potion.get("effect", {}):
```

Comment:
The `potion` attribute is expected to be a dictionary with an 'effect' key. Ensure that all potions in the `KNIGHTS` dictionary follow this structure to avoid runtime errors.